src/server/snapshot.cc

```diff
@@ -37,8 +38,80 @@ size_t SliceSnapshot::DbRecord::size() const {
   return HeapSize(value);
 }
 
+struct SliceSnapshot::TieredSerializer {
+  struct ValueBase {
+    DbIndex dbid;
+    CompactObj key;
+    time_t expire;
+  };
+
+  using PendingValue = std::pair<ValueBase, util::fb2::Future<PrimeValue>>;
+  using DelayedValue = std::pair<ValueBase, PrimeValue /* stores segment and encoding flags */>;
+
+  static const int kMaxPageAccum = 100000;
+
+  void Save(DbIndex db_indx, const PrimeKey& pk, const PrimeValue& pv, time_t expire_time) {
+    ValueBase base{db_indx, PrimeKey(pk.ToString()), expire_time};
+
+    // Delay serialization of small values to possibly find more for the same page,
+    // reducing the total number of issued io requests.
+    if (auto [offset, size] = pv.GetExternalSlice(); size < TieredStorage::kMinOccupancySize) {
+      PrimeValue pv_copy;
+      pv_copy.ImportExternal(pv);
+      size_t page = offset / tiering::kPageSize;
+      auto& entries = delayed_[page];
+      {
+        delayed_sizes_.erase({entries.size(), page});
+        entries.emplace_back(std::move(base), std::move(pv_copy));
+        delayed_sizes_.insert({entries.size(), page});
+      }
+    } else {
+      pending_.push_back(Read(std::move(base), pv));
+    }
+  }
+
+  void FlushDelayed(bool force) {
+    // Flush pages with most records accumulated first, or all, if forced.
+    // It's enough just to issue reads, because they are collapsed by the tiered storage internally
+    while ((force && !delayed_.empty()) || delayed_.size() > kMaxPageAccum) {
```

Comment:
We need to choose some kind of limit here... by the number of pages? by the total sum of enties? 🤷🏻‍♂️ We have to take into account that once we flush all of those, we can get a memory spike